peer/record.go

```diff
@@ -0,0 +1,188 @@
+package peer
+
+import (
+	"errors"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/peer/pb"
+	"github.com/libp2p/go-libp2p-core/record"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// The domain string used for peer records contained in a SignedEnvelope.
+const PeerRecordEnvelopeDomain = "libp2p-peer-record"
+
+// The type hint used to identify peer records in a SignedEnvelope.
+// TODO: register multicodec
+var PeerRecordEnvelopePayloadType = []byte("/libp2p/peer-record")
+
+// ErrPeerIdMismatch is returned when attempting to sign a PeerRecord using a key that
+// does not match the PeerID contained in the record.
+var ErrPeerIdMismatch = errors.New("unable to sign peer record: key does not match record.PeerID")
+
+// PeerRecord contains information that is broadly useful to share with other peers,
+// either through a direct exchange (as in the libp2p identify protocol), or through
+// a Peer Routing provider, such as a DHT.
+//
+// Currently, a PeerRecord contains the public listen addresses for a peer, but this
+// is expected to expand to include other information in the future.
+//
+// PeerRecords are intended to be signed by the peer they describe, and there are no
+// public methods for marshalling and unmarshalling unsigned PeerRecords.
+//
+// To share a PeerRecord, first call Sign to wrap the record in a SignedEnvelope
+// and sign it with the local peer's private key:
+//
+//     rec := NewPeerRecord(myPeerId, myAddrs)
+//     envelope, err := rec.Sign(myPrivateKey)
+//
+// The resulting record.SignedEnvelope can be marshalled to a []byte and shared
+// publicly. As a convenience, the MarshalSigned method will produce the
+// SignedEnvelope and marshal it to a []byte in one go:
+//
+//     rec := NewPeerRecord(myPeerId, myAddrs)
+//     recordBytes, err := rec.MarshalSigned(myPrivateKey)
+//
+// To validate and unmarshal a signed PeerRecord from a remote peer, use the
+// UnmarshalSignedPeerRecord function:
+//
+//    rec, envelope, err := UnmarshalSignedPeerRecord(recordBytes)
```

Comment:
Usually in go, you unmarshal into an existing (empty) struct. This allows you to control memory allocation and potentially reuse structs.

So this would look like:

```go
var rec *PeerRecord
err := rec.Unmarshal(bytes)
```

I also think that unsealing the envelope should return the bytes, and the type of the record. And the caller should be responsible for unmarshalling the record into the concrete backing struct.

Reply:
Sounds good 👍

Looking at it now, I agree that this does too much, and we should let the caller deal with unwrapping the envelope. I'll change this to accept an empty `*PeerRecord`.